transcode/transcoder_caching.go

```diff
@@ -64,6 +68,41 @@ func (t *CachingTranscoder) Transcode(ctx context.Context, profile Profile, in s
 	return nil
 }
 
+func (t *CachingTranscoder) CacheEject() {
+	// Delete LRU cache files that exceed size limit. Use last modified time.
+	type file struct {
+		path string
+		info os.FileInfo
+	}
+
+	var files []file
+	var total int64 = 0
+
+	_ = filepath.Walk(t.cachePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			files = append(files, file{path, info})
+			total += info.Size()
+		}
+		return nil
+	})
+
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].info.ModTime().Before(files[j].info.ModTime())
+	})
+
+	for total > int64(t.limitMb)*1024*1024 {
+		curFile := files[0]
+		files = files[1:]
+		total -= curFile.info.Size()
+		unlock := t.locks.Lock(curFile.path)
```

Comment:
the t.locks is a keyedMutex but the keys are profile strings (ffmpeg commands with args and paths) not paths. so this has little effect i think

maybe we should have a global clean lock which the normal transcode thing respects too?

something like

```
type CachingTranscoder struct {
	...
	locks      keyedMutex
    cleanLock  sync.RWMutex
}

...

func (t *CachingTranscoder) Transcode(...) error {
    ...

    t.cleanLock.RLock()
    defer t.cleanLock.RUnlock()
    
    ...

	key := cacheKey(name, args)
	unlock := t.locks.Lock(key)
	defer unlock()
}

func (t *CachingTranscoder) CacheEject() error {
    t.cleanLock.Lock()
    defer t.cleanLock.Unlock()
}
```

that way the cleaning process locks the entire cache, while multiple goroutines looking to read the cache work fine, just not while we're cleaning

wdyt?

Reply:
Good call and my apologies. Done.